Fix uploader creation on SIP upload

jraddaoui opened the pull request:
Use a single transaction to find or create the user and create the SIP. Reuse the User datatype for the SIP Uploader to be able to pass the OIDC fields on SIP creation. Also, check user claims before streaming the upload and rename SIP edge in the DB schema from to uploader.

Refs #1291.

djjuhasz commented on internal/persistence/ent/client/sip.go:
```diff
@@ -51,20 +52,33 @@ func (c *client) CreateSIP(ctx context.Context, s *datatypes.SIP) error {
-		u, err := tx.User.Query().Where(user.UUID(s.Uploader.UUID)).Only(ctx)
+	// If Uploader is set, find or create the user and link it to the SIP.
+	if s.Uploader != nil {
+		// Try to find the user by OIDC Iss and Sub.
```
This is a fairly long conditional block with nested conditionals — I think the code would be easier to read if the logic to get or create the uploader was moved to its own function.

jraddaoui replied:
Done!

djjuhasz commented on internal/persistence/ent/client/sip.go:
```diff
@@ -51,20 +52,33 @@ func (c *client) CreateSIP(ctx context.Context, s *datatypes.SIP) error {
+				return rollback(tx, err)
+			}
+
+			// User does not exist, create a new one. Allow the UUID to be
```
Nitpick: "The user does not exist, so create a new one" or similar to be a proper sentence.

djjuhasz commented on internal/persistence/ent/client/sip.go:
```diff
@@ -51,20 +52,33 @@ func (c *client) CreateSIP(ctx context.Context, s *datatypes.SIP) error {
+			// User does not exist, create a new one. Allow the UUID to be
+			// set by the caller, or generate a new one if not provided.
+			if s.Uploader.UUID == uuid.Nil {
+				s.Uploader.UUID = uuid.New()
```
I'm not a fan of generating a UUID here — it's non-deterministic and not tested. I'd prefer to generate the UUID in the `goaWrapper` caller where we have an injected RNG that can be deterministic for tests.

jraddaoui replied:
Okay, I made it a required field if the uploader is set. And I did the same for the OIDC fields, as they are required in readOIDCUser.

djjuhasz commented on internal/persistence/ent/client/sip.go:
```diff
@@ -51,20 +52,33 @@ func (c *client) CreateSIP(ctx context.Context, s *datatypes.SIP) error {
-			Email: u.Email,
-			Name:  u.Name,
+		// Get and set the uploader ID on the SIP.
+		uID, err := tx.User.Query().Where(user.UUID(s.Uploader.UUID)).OnlyID(ctx)
```
I think it would be better to return the user ID (PK) from `readOIDCUser()` then to do a second database query for the same user.  Alternately you could use a direct ent query to check if the user exists in the first place.

jraddaoui replied:
We recently talk about using UUIDs in the internal service interfaces too, and the `User` struct from the datatypes package doesn't even include the DB ID, something that I was planning to do in the other types when the service interfaces is fully changed. And we don't get the DB ID from the user creation either.

I was trying to use the existing functions/process as much as possible but, now that you say this, I could return the DB ID from the private functions as another var and just ignore it from the public functions. I will do that ;)

djjuhasz commented on internal/persistence/ent/client/sip.go:
```diff
@@ -76,14 +90,11 @@ func (c *client) CreateSIP(ctx context.Context, s *datatypes.SIP) error {
-	// Update SIP with DB data, to get generated values (e.g. ID).
+	// Update SIP with DB data, to get generated values (e.g. ID). Manually set
+	// the uploader data because the dbs result doesn't include the user edge.
+	uploader := s.Uploader
```
`s.Uploader` is not getting updated from the database when you create a new user, so I think `uploader` will be missing the `CreatedAt` value, and could have the wrong `Email` and `Name` if the `s.Uploader` values are different than the database values.

jraddaoui replied:
I think it's updated. If we get the user from `readOIDCUser` it's transformed [here](https://github.com/artefactual-sdps/enduro/blob/dev/issue-1291-create-user-tx/internal/persistence/ent/client/user.go#L90) and assigned to the `s.Uploader` [here](https://github.com/artefactual-sdps/enduro/blob/dev/issue-1291-create-user-tx/internal/persistence/ent/client/sip.go#L73), if we create a new user it's transformed [here](https://github.com/artefactual-sdps/enduro/blob/dev/issue-1291-create-user-tx/internal/persistence/ent/client/user.go#L46) where the passed `s.Uploader` is also being updated.

djjuhasz replied:
Oh, you're right.  I forgot that `createUser()` mutates the passed user. :+1:

djjuhasz commented on internal/persistence/ent/client/user.go:
```diff
@@ -15,12 +15,16 @@ import (
+	return createUser(ctx, c.ent, u)
+}
+
+func createUser(ctx context.Context, entc *db.Client, u *datatypes.User) error {
```
It wasn't obvious to me that the separate `createUser` and `readUser` functions were to allow passing in a database transaction client — I think it would be helpful to add a comment to explain this.

djjuhasz commented on internal/persistence/ent/schema/sip.go:
```diff
@@ -66,7 +66,7 @@ func (SIP) Edges() []ent.Edge {
 		edge.To("workflows", Workflow.Type).
 			Annotations(entsql.OnDelete(entsql.Cascade)),
-		edge.From("user", User.Type).
+		edge.From("uploader", User.Type).
```
This is a nice little change for readability in the code. :+1:

djjuhasz commented on internal/datatypes/sip.go:
```diff
@@ -42,7 +36,7 @@ type SIP struct {
 
 	// Uploader is the user that uploaded the SIP.
-	Uploader *Uploader
+	Uploader *User
```
Dropping the `Uploader` type works better than I expected. :+1:

djjuhasz reviewed (changes requested):
Looking good @jraddaoui, I just noticed a few potential problems when linking an uploader to a SIP.

jraddaoui commented:
Thanks @djjuhasz! Ready for another look.

djjuhasz commented on internal/persistence/ent/client/user.go:
```diff
@@ -15,13 +16,16 @@ import (
 
-func createUser(ctx context.Context, entc *db.Client, u *datatypes.User) error {
+// createUser creates and persists a new user.
+// It allows to pass an Ent Client that binds to a running transaction.
```
Nitpick: "It allows passing an Ent Client..."

djjuhasz commented on internal/persistence/ent/client/user.go:
```diff
@@ -66,16 +70,19 @@ func (c *client) ReadUser(ctx context.Context, id uuid.UUID) (*datatypes.User, e
 
-func readOIDCUser(ctx context.Context, entc *db.Client, iss, sub string) (*datatypes.User, error) {
+// readOIDCUser retrieves a user by OIDC issuer and subject.
+// It allows to pass an Ent Client that binds to a running transaction.
```
Here too.

djjuhasz reviewed (approved):
LGTM! Just one minor grammar issue x 2. :)

jraddaoui committed:
Fix uploader creation on SIP upload

Use a single transaction to find or create the user and create the SIP.
Reuse the User datatype for the SIP Uploader to be able to pass the
OIDC fields on SIP creation. Also, check user claims before streaming
the upload and rename SIP edge in the DB schema from to uploader.
